pkg/hubble/observer/utils.go

```diff
@@ -0,0 +1,79 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Hubble
+
+package observer
+
+import (
+	"fmt"
+	"strings"
+
+	flowpb "github.com/cilium/cilium/api/v1/flow"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/known/fieldmaskpb"
+)
+
+type filter map[string]filter
+
+func createFilter(fm *fieldmaskpb.FieldMask) (filter, error) {
+	if fm == nil {
+		return nil, nil
+	}
+	if !fm.IsValid(&flowpb.Flow{}) {
+		return nil, fmt.Errorf("invalid fieldmask")
+	}
+	fm.Normalize()
+	f := filter(make(map[string]filter))
+	for _, path := range fm.GetPaths() {
+		f.add(path)
+	}
+	return f, nil
+}
+
+func (f filter) add(path string) {
+	sub := strings.SplitN(path, ".", 2)
+	switch len(sub) {
+	case 1:
+		// ["src.id", "src"] doesn't occur due to fm.Normalize()
+		f[path] = make(map[string]filter)
+	case 2:
+		if _, ok := f[sub[0]]; !ok {
+			f[sub[0]] = make(map[string]filter)
```

Comment:
While it won't be possible to avoid allocating this map (as we use it right after), it is possible to replace previous allocation which I did. There is also the possibility to not build any structures (by processing `copy` with fieldmask), but in my opinion it would decrease performance and code readability.

I added a more complex example of `l4.TCP.source_port`. It works